src/Oscoin/Storage/Ledger.hs

```diff
@@ -0,0 +1,282 @@
+--
+-- @initialState@ is the state that the transactions of the
+-- genesisblock are applied to.
+newMemoryIO
+    :: ( MonadIO m
+       , Crypto.Hashable crypto state
+       , Crypto.Hashable crypto tx
+       )
+    => Evaluator state tx output
+    -> Chrono.NewestFirst NonEmpty (Block crypto tx (Sealed crypto seal))
+    -> state
+    -> IO (Ledger crypto seal tx output state m)
+newMemoryIO ledgerEval blks initialState = do
+    blockStore <- newBlockStoreIO blks
+    newFromBlockStore ledgerEval blockStore initialState
+
+
+-- | Same as 'newMemoryIO' with just @genesisBlk@ in the blockchain.
+newMemoryFromGenesisIO
+    :: ( MonadIO m
+       , Crypto.Hashable crypto state
+       , Crypto.Hashable crypto tx
+       )
+    => Evaluator state tx output
+    -> Block crypto tx (Sealed crypto seal)
+    -> state
+    -> IO (Ledger crypto seal tx output state m)
+newMemoryFromGenesisIO ledgerEval genesisBlk initialState =
+    let blks = Chrono.NewestFirst $ genesisBlk NonEmpty.:| []
+    in newMemoryIO ledgerEval blks initialState
+
+
+-- | Create a ledger backed by the given block store and in-memory
+-- storage of state and receipts.
+--
+-- @initialState@ is the state that the transactions of the
+-- genesisblock are applied to.
+newFromBlockStore
+    :: ( MonadIO m
+       , Crypto.Hashable crypto state
+       , Crypto.Hashable crypto tx
+       )
+    => Evaluator state tx output
+    -> BlockStore crypto tx seal IO
+    -> state
+    -> IO (Ledger crypto seal tx output state m)
+newFromBlockStore ledgerEval blockStore initialState = do
+    receiptStore <- ReceiptStore.newReceiptStoreIO
+    stateStore <- newHashStoreIO
+    storeHashContent stateStore initialState
+
+    let ledger = Ledger{blockStore, stateStore, receiptStore, ledgerEval}
+    materialize ledger initialState
+    pure $ hoist liftIO ledger
+
+
+mkLedger
+    :: BlockStore crypto tx seal m
+    -> HashStore crypto state m
+    -- ^ State store
+    -> Evaluator state tx output
+    -> ReceiptStore.ReceiptStore crypto tx output m
+    -> Ledger crypto seal tx output state m
+mkLedger blockStore stateStore ledgerEval receiptStore =
+    Ledger
+        { blockStore
+        , ledgerEval
+        , stateStore
+        , receiptStore
+        }
+
+
+-----------------------------------------------------------
+-- * Getters
+-----------------------------------------------------------
+
+blockStoreReader :: Ledger crypto seal tx output state m -> BlockStore.BlockStoreReader crypto tx seal m
+blockStoreReader = fst . blockStore
+
+-- | Get the @depth@ most recent blocks.
+getBlocksByDepth
+    :: Ledger crypto seal tx output state m
+    -> Depth
+    -> m (Chrono.NewestFirst [] (Block crypto tx (Sealed crypto seal)))
+getBlocksByDepth ledger depth = BlockStore.getBlocksByDepth (blockStoreReader ledger) depth
+
+getTip :: Ledger crypto seal tx output state m -> m (Block crypto tx (Sealed crypto seal))
+getTip ledger = BlockStore.getTip (fst $ blockStore ledger)
```

Comment:
This always gets the tip of the best chain, correct? Remind me how that works? Ie. how the ledger is always operating on the best fork?